Feature nxp 29220 explorer xml secure content

atchertchian opened the pull request:
Draft PR on branch feature-NXP-29050-explorer-hide-current-distrib, waiting for #4133 to be merged

nuxeojenkins commented:
View issue in JIRA: [NXP-29220](https://jira.nuxeo.com/browse/NXP-29220): Improve Explorer secure XML content logics

efge commented on modules/platform/nuxeo-apidoc-server/nuxeo-apidoc-core/src/main/java/org/nuxeo/apidoc/documentation/DocumentationHelper.java:
```diff
@@ -134,15 +148,54 @@ public static String getHtml(String doc) {
+            if (res.contains(kw)) {
+                for (String pattern : List.of(
+                        // node startswith
+                        String.format("(?<start><(?<key>\\w*%s)\\s*>)[^<]*(?<end></\\w*%s>)", kw, kw),
```
I would have done the rewriting using StAX instead of ad hoc regexps...

atchertchian replied:
As discussed, i hesitated, and then changed my mind. I will see that again with https://jira.nuxeo.com/browse/NXP-28948 if you're ok with it as i will need to handle cases where XML would not be valid (while regexps would still work)

atchertchian replied:
After more discussion, will use StAX as primary parser, to handle more use cases, and fallback on regexps in case of invalid XML

kevinleturc commented on modules/platform/nuxeo-apidoc-server/nuxeo-apidoc-core/src/main/java/org/nuxeo/apidoc/documentation/DocumentationHelper.java:
```diff
@@ -134,15 +148,54 @@ public static String getHtml(String doc) {
+
+    protected static List<String> getKeywordList(String property, List<String> defaultValue) {
+        ConfigurationService cs = Framework.getService(ConfigurationService.class);
+        String value = cs.getString(property, null);
+        if (value == null) {
+            return defaultValue;
+        }
+        return List.of(value.split("\\s*,[,\\s]*"));
```
Something like below should be used as the ConfigurationService APIs with default value are not designed to pass null.
```
return cs.getString(property).map(v -> v.split("\\s*,[,\\s]*")).map(List::of).orElse(defaultValue);
```

atchertchian commented:
T&P running at https://qa.nuxeo.org/jenkins/job/TestAndPush/job/ondemand-testandpush-atchertchian-master/245/

kevinleturc commented on modules/platform/nuxeo-apidoc-server/nuxeo-apidoc-core/src/main/java/org/nuxeo/apidoc/documentation/SecureXMLHelper.java:
```diff
@@ -0,0 +1,209 @@
+
+    public static String secureStAX(String xml, List<String> keywords, List<String> whitelist)
+            throws XMLStreamException {
+        // System.out.println("xml = " + xml);
```
Wanted?

atchertchian replied:
No, thanks

kevinleturc commented on modules/platform/nuxeo-apidoc-server/nuxeo-apidoc-core/src/main/java/org/nuxeo/apidoc/documentation/SecureXMLHelper.java:
```diff
@@ -0,0 +1,209 @@
+            throws XMLStreamException {
+        // System.out.println("xml = " + xml);
+
+        InputStream stream = new ByteArrayInputStream(xml.getBytes());
+        XMLEventReader reader = inputFactory.createXMLEventReader(stream);
+        OutputStream output = new ByteArrayOutputStream();
+        XMLEventWriter writer = outputFactory.createXMLEventWriter(output);
```
The `stream` InputStream doesn't seem to be closed, could we use try-with-resources to do that?

atchertchian replied:
As discussed, byte array input streams do not need to be closed, however could be a good practice, wdyt @efge please?

efge replied:
@atchertchian indeed `ByteArrayInputStream` doesn't need to be closed, the GC will deal with it, however most tools and IDEs won't recognize that the `InputStream` you use later on comes from a `ByteArrayInputStream` and will flag it as a leaked resource. So you might as well use try-with-resources there too.

atchertchian commented:
Reworked the commits, fixed tests by avoiding changing xml at all when keywords not found in it + addressed feedback -> new T&P https://qa.nuxeo.org/jenkins/job/TestAndPush/job/ondemand-testandpush-atchertchian-master/247

efge commented on modules/platform/nuxeo-apidoc-server/nuxeo-apidoc-core/src/main/java/org/nuxeo/apidoc/documentation/SecureXMLHelper.java:
```diff
@@ -0,0 +1,220 @@
+        if (!isChangeNeeded(xml, keywords)) {
+            return xml;
+        }
+        try (InputStream stream = new ByteArrayInputStream(xml.getBytes());
```
`getBytes(UTF_8)`

efge commented on modules/platform/nuxeo-apidoc-server/nuxeo-apidoc-core/src/main/java/org/nuxeo/apidoc/documentation/SecureXMLHelper.java:
```diff
@@ -0,0 +1,220 @@
+
+    protected static final String SECRET_VALUE = "********";
+
+    protected static final XMLInputFactory inputFactory = XMLInputFactory.newInstance();
```
constants -> all caps

efge commented on modules/platform/nuxeo-apidoc-server/nuxeo-apidoc-core/src/main/java/org/nuxeo/apidoc/documentation/SecureXMLHelper.java:
```diff
@@ -0,0 +1,220 @@
+            writer.close();
+
+            String all = output.toString();
+            return all;
```
Could be just one line

efge commented on modules/platform/nuxeo-apidoc-server/nuxeo-apidoc-core/src/main/java/org/nuxeo/apidoc/documentation/SecureXMLHelper.java:
```diff
@@ -0,0 +1,220 @@
+            writer.flush();
+            writer.close();
+
+            String all = output.toString();
```
Define `ByteArrayOutputStream output` above, and use `output.toString(UTF_8)` to convert to string. Charset is important.

efge commented on modules/platform/nuxeo-apidoc-server/nuxeo-apidoc-core/src/main/java/org/nuxeo/apidoc/documentation/SecureXMLHelper.java:
```diff
@@ -0,0 +1,220 @@
+                        // node endswith
+                        String.format("(?<start><(?<key>%s\\w*)\\s*>)[^<]*(?<end></%s\\w*>)", kw, kw),
+                        // attributes startswith
+                        String.format("(?<start>(?<key>\\w*%s)=\")[^\"]*(?<end>\")", kw),
```
Could you also add the variants with a `'` to delimit attribute values?

atchertchian commented:
https://qa.nuxeo.org/jenkins/job/TestAndPush/job/ondemand-testandpush-atchertchian-master/248/

NourNuxeo commented on modules/platform/nuxeo-apidoc-server/nuxeo-apidoc-repo/src/test/java/org/nuxeo/apidoc/test/TestSecureXML.java:
```diff
@@ -15,32 +15,174 @@
+                + " <secretKey>${nuxeo.aws.secretKey}</secretKey>\n" //
+                + " <option name=\"apiKey\">${metrics.datadog.apiKey}</option>\n" //
+                + " <passwordField>password</passwordField>\n" //
+                + " <passwordHashAlgorithm>SSHA</passwordHashAlgorithm>";
```
It seems to me there is a case that is not tested ?
`<password></password>`
should also be replaced with
`<password>********</password>`as you do have `[^<]*` in your regex.
An empty password wouldn't be hard to find anyway

atchertchian replied:
It's tested here: https://github.com/nuxeo/nuxeo/pull/4168/files#diff-172201976b34d3f53d4238f8bd458f38R208
I assumed an empty password did not need to be replaced, but indeed, maybe better to make it a complete mystery (?)

NourNuxeo replied:
Sorry but the anchor doesn't seem to get me to the right place ?
About misteries, I think putting stars doesn't give away the password is empty

atchertchian replied:
Anchor seems broken because it's a large diff https://github.com/nuxeo/nuxeo/pull/4168/commits/ee3f00ccd1519b8dc0041c65f9fe91fabd322a17#diff-172201976b34d3f53d4238f8bd458f38R215

Anyway i'll do the change

atchertchian commented:
https://qa.nuxeo.org/jenkins/job/TestAndPush/job/ondemand-testandpush-atchertchian-master/249/

efge commented on modules/platform/nuxeo-apidoc-server/nuxeo-apidoc-core/src/main/java/org/nuxeo/apidoc/documentation/SecureXMLHelper.java:
```diff
@@ -0,0 +1,276 @@
+            writer.close();
+
+            return output.toString(StandardCharsets.UTF_8);
+        } catch (
+
+        IOException e) {
```
Weird indent here

atchertchian commented:
Last status is visible in unit tests, note some extra pseudo replacements of empty tags in last commit. I could replace content only on match on attribute name and element name, but if so i'd have a list of all use cases to cover from the beginning, and i'd also rather have an XMap annotation for all sensitive content, it would clearly make things cleaner here.
